### ISSUE TITLE
Unable to use client certificate authentication with Solace queue provider

### ISSUE THREAD

**Issue by jfeyzeau** [jmstoolbox/jmstoolbox#147]
## Specifications
- JMSToolBox version: v6.0.0
- OS (Name, version, 32/64bits): Windows 11 x64
- Q provider (kind, version): Solace (default version)

## Expected Behavior
Connect to host using client certificate authentication scheme when the properties `ssl_key_store` and `ssl_key_store_password` are set.

## Actual Behavior
JMSTollBox does not try to open the keystore and attempt a basic authentication instead.

## Steps to Reproduce the Problem

  1. Open attached zip archive, it contains:
      + The full log files under ./logs
      + A docker-compose.yaml file to spin up a solace test container (under ./docker)
      + Some sample certificates and private keys for the TLS config (under ./pki)
      + Screenshots of the configuration required on solace management interface
  2. Deploy a test solace instance using attached docker-compose.yaml file and sample server.pem:
      + Change the path of the secrets folder mount in the `docker-compose.yaml` file to reflect where you extracted the archive: ```line 10: - "/path-where-you-unzipped-the-archive/JMSToolBox-issue/docker/secrets:/run/secrets"``` 
      + From inside the docker folder, simply run a `docker compose up`
  3. Configure solace to
      + Add Client Certificate Authority root CA (cf screenshot Solace-1)
      + Add a user "client-cert-auth" (cf screenshot Solace-2 & Solace-3)
      + Accept only client certificate authentication (cf screenshot Solace-4)
  4. Open JMSToolBox, configure a Solace Session with the SSL Keystore properties setup (See attached config.xml)
  5. Try to connect, I get an error stating that `403: Basic Authentication Is Shutdown` which is indeed the case, but I couldn't find any way to make JMSToolBox try to use the configured keystore.

I attached a sample keystore in pkcs12 format (pki/keystore.p12) with a private key and client cert signed by a local root ca (pki/localCA.crt). Password for keystore and private key is the same: "change_me".

## config.xml
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<config>
    <qManagerDef id="org.titou10.jtb.qm.solace" name="Solace"/>
    <sessionDef qManagerDef="org.titou10.jtb.qm.solace" name="solace-local">
        <host>tcps://127.0.0.1</host>
        <port>55443</port>
        <userid>client-cert-auth</userid>
        <promptForCredentials>false</promptForCredentials>
        <properties>
            <property name="VPN" value="default" kind="STRING"/>
            <property name="browser_timeout" value="250" kind="INT"/>
            <property name="mgmt_password" value="{##}/2g3PB5jdOZz9RzU9NTEHw==" kind="STRING"/>
            <property name="mgmt_url" value="https://127.0.0.1:1943" kind="STRING"/>
            <property name="mgmt_username" value="admin" kind="STRING"/>
            <property name="ssl_key_store" value="C:\JMSToolBox\keystore.p12" kind="STRING"/>
            <property name="ssl_key_store_format" value="pkcs12" kind="STRING"/>
            <property name="ssl_key_store_password" value="{##}zW2enZs5CBhd9UiZpS6XxA==" kind="STRING"/>
            <property name="ssl_validate_certificate" value="false" kind="BOOLEAN"/>
        </properties>
    </sessionDef>
</config>
```

## Logs
```
2023-02-22 17:20:17:601 INFO  [o.t.j.c.ConfigManager       ] ***************************************************
2023-02-22 17:20:17:601 INFO  [o.t.j.c.ConfigManager       ] * JMSToolBox v6.0.0 successfully initialized with:
2023-02-22 17:20:17:602 INFO  [o.t.j.c.ConfigManager       ] * -  16 installed plugins
2023-02-22 17:20:17:602 INFO  [o.t.j.c.ConfigManager       ] * -  14 running plugins
2023-02-22 17:20:17:602 INFO  [o.t.j.c.ConfigManager       ] * -   1 external connector plugins
2023-02-22 17:20:17:602 INFO  [o.t.j.c.ConfigManager       ] * -   1 QManagersDefs
2023-02-22 17:20:17:603 INFO  [o.t.j.c.ConfigManager       ] * -   1 sessions
2023-02-22 17:20:17:603 INFO  [o.t.j.c.ConfigManager       ] * -   0 scripts
2023-02-22 17:20:17:603 INFO  [o.t.j.c.ConfigManager       ] * -   6 variables
2023-02-22 17:20:17:603 INFO  [o.t.j.c.ConfigManager       ] * -   6 visualizers
2023-02-22 17:20:17:604 INFO  [o.t.j.c.ConfigManager       ] * -   1 templates root folders
2023-02-22 17:20:17:604 INFO  [o.t.j.c.ConfigManager       ] * -   1 columns sets
2023-02-22 17:20:17:604 INFO  [o.t.j.c.ConfigManager       ] * -   3 session types
2023-02-22 17:20:17:604 INFO  [o.t.j.c.ConfigManager       ] *
2023-02-22 17:20:17:605 INFO  [o.t.j.c.ConfigManager       ] * System Information:
2023-02-22 17:20:17:605 INFO  [o.t.j.c.ConfigManager       ] * - OS   : Name=Windows 11 Version=10.0 Arch=amd64
2023-02-22 17:20:17:605 INFO  [o.t.j.c.ConfigManager       ] * - Java : 17.0.5 (17.0.5+8-LTS) Azul Systems, Inc. 
2023-02-22 17:20:17:605 INFO  [o.t.j.c.ConfigManager       ] ***************************************************
2023-02-22 17:20:17:605 DEBUG [o.t.j.c.ConfigManager       ] 
2023-02-22 17:20:18:834 DEBUG [p.JTBSessionsBrowserViewPart] createControls
2023-02-22 17:20:18:835 DEBUG [p.JTBSessionsBrowserViewPart] buildSessionList
2023-02-22 17:20:19:572 DEBUG [o.t.j.c.MainWindowCloseAddon] register MainWindowCloseAddon Add On
2023-02-22 17:20:25:990 DEBUG [.t.j.h.SessionConnectHandler] execute. Selection : NodeJTBSession [solace-local]
2023-02-22 17:20:25:991 DEBUG [o.t.j.j.m.JTBConnection     ] connect 'JTBConnection [jtbSessionClientType=GUI, isConnected()=false]'
2023-02-22 17:20:25:992 INFO  [o.t.j.q.s.SolaceQManager    ] connecting to solace-local - JMSToolBox-1677082818908
2023-02-22 17:20:29:382 ERROR [o.t.j.u.JTBStatusReporter   ] Connect unsuccessful
com.solacesystems.jcsmp.JCSMPErrorResponseException: 403: Basic Authentication Is Shutdown
	at com.solacesystems.jcsmp.protocol.impl.TcpChannel.executePostOnce(TcpChannel.java:235)
	at com.solacesystems.jcsmp.protocol.impl.ChannelOpStrategyClient.performOpen(ChannelOpStrategyClient.java:99)
	at com.solacesystems.jcsmp.protocol.impl.TcpClientChannel.performOpenSingle(TcpClientChannel.java:426)
	at com.solacesystems.jcsmp.protocol.impl.TcpClientChannel.access$800(TcpClientChannel.java:122)
	at com.solacesystems.jcsmp.protocol.impl.TcpClientChannel$ClientChannelConnect.call(TcpClientChannel.java:2595)
	at com.solacesystems.jcsmp.protocol.impl.TcpClientChannel.open(TcpClientChannel.java:402)
	at com.solacesystems.jcsmp.impl.JCSMPBasicSession.sniffRouter(JCSMPBasicSession.java:407)
	at com.solacesystems.jcsmp.impl.JCSMPBasicSession.connect(JCSMPBasicSession.java:1264)
	at com.solacesystems.jms.SolConnection.<init>(SolConnection.java:167)
	at com.solacesystems.jms.SolConnection.<init>(SolConnection.java:97)
	at com.solacesystems.jms.SolConnectionFactoryImpl.createConnection(SolConnectionFactoryImpl.java:114)
	at org.titou10.jtb.qm.solace.SolaceQManager.connect(SolaceQManager.java:291)
        ...
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
```
See full logs in zip (`logs/jmstoolbox.log` and `logs/.log`    )

## Possibly interesting additional information
I am not really familiar with java, but browsing your files, I found the file where you instantiate the SolSession of the lib provided by Solace.
[https://github.com/jmstoolbox/jmstoolbox/blob/dev/org.titou10.jtb.qm.solace/src/org/titou10/jtb/qm/solace/SolaceQManager.java](https://github.com/jmstoolbox/jmstoolbox/blob/dev/org.titou10.jtb.qm.solace/src/org/titou10/jtb/qm/solace/SolaceQManager.java)
From line 234 to 291:
```
      // JMS Connections

      SolConnectionFactory cf = SolJmsUtility.createConnectionFactory();
      cf.setDirectTransport(false);

      cf.setHost(sessionDef.getHost());
      cf.setPort(sessionDef.getPort());

      cf.setVPN(vpn);
      cf.setUsername(sessionDef.getActiveUserid());
      cf.setPassword(sessionDef.getActivePassword());
      cf.setBrowserTimeoutInMS(browserTimeout
);

      if (sslCipherSuite != null) {
         cf.setSSLCipherSuites(sslCipherSuite);
      }
      if (sslConnectionDowngradeTo != null) {
         cf.setSSLConnectionDowngradeTo(sslConnectionDowngradeTo);
      }
      if (sslExcludedProtocols != null) {
         cf.setSSLExcludedProtocols(sslExcludedProtocols);
      }
      if (sslKeyStore != null) {
         cf.setSSLKeyStore(sslKeyStore);
      }
      if (sslKeyStoreFormat != null) {
         cf.setSSLKeyStoreFormat(sslKeyStoreFormat);
      }
      if (sslKeyStorePassword != null) {
         cf.setSSLKeyStorePassword(sslKeyStorePassword);
      }
      if (sslPrivateKeyAlias != null) {
         cf.setSSLPrivateKeyAlias(sslPrivateKeyAlias);
      }
      if (sslPrivateKeyPassword != null) {
         cf.setSSLPrivateKeyPassword(sslPrivateKeyPassword);
      }
      if (sslProtocol != null) {
         cf.setSSLProtocol(sslProtocol);
      }
      if (sslTrustStore != null) {
         cf.setSSLTrustStore(sslTrustStore);
      }
      if (sslTrustStoreFormat != null) {
         cf.setSSLTrustStoreFormat(sslTrustStoreFormat);
      }
      if (sslTrustStorePassword != null) {
         cf.setSSLTrustStorePassword(sslTrustStorePassword);
      }
      if (sslTrustedCommonNameList != null) {
         cf.setSSLTrustedCommonNameList(sslTrustedCommonNameList);
      }
      if (sslValidateCertificate != null) {
         cf.setSSLValidateCertificate(Boolean.parseBoolean(sslValidateCertificate));
      }
      if (sslValidateCertificateDate != null) {
         cf.setSSLValidateCertificateDate(Boolean.parseBoolean(sslValidateCertificateDate));
      }

      Connection jmsConnection = cf.createConnection();
```

Browsing [the doc from Solace for the SolConnectionFactory interface](https://docs.solace.com/API-Developer-Online-Ref-Documentation/jms/javax/jms/ConnectionFactory.html) I can see that there is a [setAuthenticationScheme](https://docs.solace.com/API-Developer-Online-Ref-Documentation/jms/com/solacesystems/jms/SolConnectionFactory.html#setAuthenticationScheme--) method that I beleive should be set depending on the property found in the defined session:
>     void setAuthenticationScheme([String](http://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true) scheme)
> 
>     This property specifies the authentication scheme.
> 
>     Values are one of
>         [SupportedProperty.AUTHENTICATION_SCHEME_BASIC](https://docs.solace.com/API-Developer-Online-Ref-Documentation/jms/com/solacesystems/jms/SupportedProperty.html#AUTHENTICATION_SCHEME_BASIC) (username/password authentication)
>         [SupportedProperty.AUTHENTICATION_SCHEME_CLIENT_CERTIFICATE](https://docs.solace.com/API-Developer-Online-Ref-Documentation/jms/com/solacesystems/jms/SupportedProperty.html#AUTHENTICATION_SCHEME_CLIENT_CERTIFICATE) (Client Certificate based authentication)
>         [SupportedProperty.AUTHENTICATION_SCHEME_GSS_KRB](https://docs.solace.com/API-Developer-Online-Ref-Documentation/jms/com/solacesystems/jms/SupportedProperty.html#AUTHENTICATION_SCHEME_GSS_KRB) (Kerberos based authentication)
>         [SupportedProperty.AUTHENTICATION_SCHEME_OAUTH2](https://docs.solace.com/API-Developer-Online-Ref-Documentation/jms/com/solacesystems/jms/SupportedProperty.html#AUTHENTICATION_SCHEME_OAUTH2) (OAuth 2.0 based authentication)
> 
>     Default: [SupportedProperty.AUTHENTICATION_SCHEME_BASIC](https://docs.solace.com/API-Developer-Online-Ref-Documentation/jms/com/solacesystems/jms/SupportedProperty.html#AUTHENTICATION_SCHEME_BASIC)
> 
>     Parameters:
>         scheme - The authentication scheme

The default for this property being AUTHENTICATION_SCHEME_BASIC and since I couldn't find another place where you set this property to AUTHENTICATION_SCHEME_CLIENT_CERTIFICATE, I think it explains why my keystore is never read.

Maybe dong some kind of condition on the fact that "sslKeyStore" is set or not and setting the authentication accordingly could do the trick. 

Also, thanks for your software, really appreciate the effort, we use it at work and its really helpfull! 
Hope this report can be help.

[JMSToolBox-issue.zip](https://github.com/jmstoolbox/jmstoolbox/files/10812882/JMSToolBox-issue.zip)

**Comment by titou10titou10** [comment]
Weird. It looks similar to issue #146 with OpenLiberty that appeared in recent version of Eclipse/Java...
I need some time to investigate

**Comment by jfeyzeau** [comment]
Hi, thank you for fixing this.

Just a note on the change you made, I see that you added all 4 means of Authentication in the documentation (AUTHENTICATION_SCHEME_BASIC (default), AUTHENTICATION_SCHEME_GSS_KRB, AUTHENTICATION_SCHEME_CLIENT_CERTIFICATE, AUTHENTICATION_SCHEME_OAUTH2), but currently, the only 2 usable are basic and client certs, because we do not have the necessary properties to configure oauth and kerberos.

Maybe we can remove them from the doc so that people do not try to use them?

**Comment by titou10titou10** [comment]
good remark. I will change it
BTW I don't have the setup to test it... 
I don't know if you are in position to build JTB by yourself, but I'll probably publish a new version after march 15 and the release of Eclipse 2023-03 (v4.27)

**Comment by jfeyzeau** [comment]
So, I was able to connect successfully using client certificate authentication as expected, after building your code from dev branch.

Many thanks!


--

I still have an issue though, when I try to use a certificate signed with a local root CA not in the JRE default truststore (as in the test setup I provided above), I land on the same error as in [#146](https://github.com/jmstoolbox/jmstoolbox/issues/146).

Even with the property "ssl_validate_certificate" set to false, the connection fails with the same "sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target", meaning JMS still tries to validate the certificate from my Solace server.

If I overwrite the default cacerts truststore with a custom one containing my root CA with the default "changeit" password, the connection still fails but not for the same reason. This time a certification path to my server cert is successfully built, but the SAN (Subject Alternative Name) validation fails:
```
2023-03-16 00:04:17:875 DEBUG [o.t.j.q.s.SolaceQManager    ] SEMP request: https://127.0.0.1:1943/SEMP/v2/config/msgVpns/default/queues?select=queueName&count=1500 GET
2023-03-16 00:04:17:888 ERROR [o.t.j.u.JTBStatusReporter   ] Connect unsuccessful
java.security.cert.CertificateException: No subject alternative names matching IP address 127.0.0.1 found
```
This seems normal since I did not setup a DNS name for my Solace server but rather uses my localhost IP, and my custom certificate does not include 127.0.0.1 in either its CN nor SAN.
But, this means that the property "ssl_trusted_common_name_list" is still not honored for this connection.

Also if I tick the "Trust all server certificates..." preference box and use the default cacerts without my local CA, I can connect successfully.

This behaviour seems quite similar to what is described in [#132](https://github.com/jmstoolbox/jmstoolbox/issues/132), where the truststore from the connection property is not used.


So I guess one would expect the ssl property defined at the connection level to apply both for the management interface connection AND the VPN/queues connection (since its the same server, chances are that they expose the same cert), instead of only the later. 
Plus this makes the SSL validation parameters for the management interface not configurable, with only a on/off toggle in the global property of JMS.


In any case, thanks again for your efforts!
